File: UltraMesh 2d animation/Data/AnimationLibrary.swift
import Foundation
import Combine

/// A named snapshot of every animation timeline in the scene: the per-bone and
/// per-sprite clips, plus the scene-wide clip that owns constraint property and
/// draw order timelines, plus the authored constraint values those timelines
/// blend away from.
///
/// The scene clip and the setup values must travel together with the rest of
/// the snapshot: without them, switching animations would silently drop every
/// constraint and draw order key, and leave constraints stranded at whatever
/// value the outgoing animation last evaluated.
struct NamedAnimation: Identifiable, Equatable {
    let id: UUID
    var name: String
    var boneClips: [UUID: AnimationClip]
    var imageClips: [UUID: AnimationClip]
    /// Constraint property and draw order timelines.
    var sceneClip: AnimationClip
    /// Authored constraint values captured when a property was first keyed.
    var constraintSetupValues: [UUID: ConstraintSetupValues]
    var duration: Int

    init(
        id: UUID,
        name: String,
        boneClips: [UUID: AnimationClip],
        imageClips: [UUID: AnimationClip],
        sceneClip: AnimationClip = AnimationClip(name: "Scene"),
        constraintSetupValues: [UUID: ConstraintSetupValues] = [:],
        duration: Int
    ) {
        self.id = id
        self.name = name
        self.boneClips = boneClips
        self.imageClips = imageClips
        self.sceneClip = sceneClip
        self.constraintSetupValues = constraintSetupValues
        self.duration = duration
    }
}

@MainActor
final class AnimationLibrary: ObservableObject {
    @Published private(set) var animations: [NamedAnimation] = []
    @Published private(set) var activeID: UUID?

    unowned let scene: SceneManager

    init(scene: SceneManager) {
        self.scene = scene
    }

    var active: NamedAnimation? {
        animations.first(where: { $0.id == activeID })
    }

    /// Puts a saved library back, as-is.
    ///
    /// Deliberately does NOT snapshot the live clips first: the project was just
    /// loaded, so those clips ARE the active animation's, and snapshotting them
    /// over the list would overwrite what was read from the file with a copy of
    /// one of its own entries.
    func restore(animations: [NamedAnimation], activeID: UUID?) {
        self.animations = animations
        // An id that names nothing would leave the panel pointing at an entry
        // that is not there; fall back to no selection rather than a dangling one.
        self.activeID = animations.contains(where: { $0.id == activeID }) ? activeID : nil
    }

    /// Captures the current state of all bone/image animationClips into a new animation.
    func createFromCurrentState(name: String) {
        let animation = snapshotCurrent(name: name)
        // Persist current edits into the previously active animation so they aren't lost.
        if let activeID, let idx = animations.firstIndex(where: { $0.id == activeID }) {
            animations[idx] = snapshotCurrent(name: animations[idx].name, id: activeID)
        }
        animations.append(animation)
        activeID = animation.id
    }

    /// Switches active animation; saves current edits into the outgoing snapshot,
    /// then loads keyframes from the target into scene clips.
    /// No-ops (but still saves) when already on the target animation, so the scene's
    /// live clips are never overwritten with a stale library snapshot.
    func switchTo(_ id: UUID) {
        if let activeID, let idx = animations.firstIndex(where: { $0.id == activeID }) {
            animations[idx] = snapshotCurrent(name: animations[idx].name, id: activeID)
        }
        guard activeID != id else { return }
        guard let target = animations.first(where: { $0.id == id }) else { return }
        applyToScene(target)
        activeID = id
    }

    func rename(_ id: UUID, to newName: String) {
        guard let idx = animations.firstIndex(where: { $0.id == id }) else { return }
        animations[idx].name = newName
    }

    func delete(_ id: UUID) {
        animations.removeAll { $0.id == id }
        if activeID == id { activeID = animations.first?.id }
        if let next = active { applyToScene(next) }
    }

    // MARK: - Internal

    private func snapshotCurrent(name: String, id: UUID = UUID()) -> NamedAnimation {
        var bones: [UUID: AnimationClip] = [:]
        var images: [UUID: AnimationClip] = [:]
        var duration = 0
        for b in scene.skeleton.orderedBones {
            bones[b.id] = b.animationClip
            duration = max(duration, b.animationClip.durationInFrames)
        }
        for img in scene.images {
            images[img.id] = img.animationClip
            duration = max(duration, img.animationClip.durationInFrames)
        }
        let sceneClip = scene.sceneAnimationClip
        duration = max(duration, sceneClip.durationInFrames)
        return NamedAnimation(
            id: id,
            name: name,
            boneClips: bones,
            imageClips: images,
            sceneClip: sceneClip,
            constraintSetupValues: scene.constraintSetupValues,
            duration: duration
        )
    }

    private func applyToScene(_ anim: NamedAnimation) {
        for boneID in scene.skeleton.orderedBones.map(\.id) {
            let clip = anim.boneClips[boneID] ?? AnimationClip(name: anim.name)
            scene.skeleton.bones[boneID]?.animationClip = clip
        }
        for imgID in scene.images.map(\.id) {
            scene.updateImage(id: imgID) { img in
                img.animationClip = anim.imageClips[imgID] ?? AnimationClip(name: anim.name)
            }
        }
        // Restore the constraints to their authored values before installing the
        // incoming timelines, so a property animated by the outgoing animation
        // but untouched by the incoming one does not keep a stale value.
        scene.restoreAllConstraintSetupValues()
        scene.replaceSceneAnimation(
            clip: anim.sceneClip,
            constraintSetupValues: anim.constraintSetupValues
        )
        scene.setCurrentFrame(scene.currentFrame)
    }
}

File: UMeshCore/src/Animation/AnimationLibrary.cpp
#include "umeshcore/Animation/AnimationLibrary.h"

#include <algorithm>

#include "umeshcore/Editor/EditorScene.h"

namespace umeshcore {

const NamedAnimation* AnimationLibrary::active() const {
    if (!activeID_.has_value()) return nullptr;
    for (const NamedAnimation& animation : animations_) {
        if (animation.id == *activeID_) return &animation;
    }
    return nullptr;
}

void AnimationLibrary::restore(std::vector<NamedAnimation> animations, std::optional<Uuid> activeID) {
    animations_ = std::move(animations);
    // An id that names nothing would leave the panel pointing at an entry
    // that is not there.
    const bool known =
        activeID.has_value() && std::any_of(animations_.begin(), animations_.end(), [&](const NamedAnimation& a) {
            return a.id == *activeID;
        });
    activeID_ = known ? activeID : std::nullopt;
}

void AnimationLibrary::createFromCurrentState(const std::string& name) {
    NamedAnimation animation = snapshotCurrent(name);
    // Persist current edits into the previously active animation so they
    // aren't lost.
    if (activeID_.has_value()) {
        for (NamedAnimation& existing : animations_) {
            if (existing.id == *activeID_) {
                existing = snapshotCurrent(existing.name, existing.id);
                break;
            }
        }
    }
    activeID_ = animation.id;
    animations_.push_back(std::move(animation));
}

void AnimationLibrary::switchTo(Uuid id) {
    // Saves the outgoing animation even when the target is already active
    // -- see this file's header.
    if (activeID_.has_value()) {
        for (NamedAnimation& existing : animations_) {
            if (existing.id == *activeID_) {
                existing = snapshotCurrent(existing.name, existing.id);
                break;
            }
        }
    }
    if (activeID_.has_value() && *activeID_ == id) return;

    const NamedAnimation* target = nullptr;
    for (const NamedAnimation& animation : animations_) {
        if (animation.id == id) {
            target = &animation;
            break;
        }
    }
    if (target == nullptr) return;

    applyToScene(*target);
    activeID_ = id;
}

void AnimationLibrary::rename(Uuid id, const std::string& newName) {
    for (NamedAnimation& animation : animations_) {
        if (animation.id == id) {
            animation.name = newName;
            return;
        }
    }
}

void AnimationLibrary::remove(Uuid id) {
    animations_.erase(
        std::remove_if(
            animations_.begin(), animations_.end(), [&](const NamedAnimation& a) { return a.id == id; }),
        animations_.end());
    if (activeID_.has_value() && *activeID_ == id) {
        activeID_ = animations_.empty() ? std::nullopt : std::optional<Uuid>(animations_.front().id);
    }
    if (const NamedAnimation* next = active()) applyToScene(*next);
}

NamedAnimation AnimationLibrary::snapshotCurrent(const std::string& name, std::optional<Uuid> id) const {
    NamedAnimation animation;
    if (id.has_value()) animation.id = *id;
    animation.name = name;

    int duration = 0;
    for (const Bone& bone : scene_.skeleton.orderedBones()) {
        animation.boneClips[bone.id] = bone.animationClip;
        duration = std::max(duration, bone.animationClip.durationInFrames);
    }
    for (const SceneImage& image : scene_.images) {
        animation.imageClips[image.id] = image.animationClip;
        duration = std::max(duration, image.animationClip.durationInFrames);
    }
    animation.sceneClip = scene_.sceneAnimationClip;
    duration = std::max(duration, animation.sceneClip.durationInFrames);

    animation.constraintSetupValues = scene_.constraintSetupValues;
    animation.duration = duration;
    return animation;
}

void AnimationLibrary::applyToScene(const NamedAnimation& animation) {
    for (const Bone& bone : scene_.skeleton.orderedBones()) {
        const auto it = animation.boneClips.find(bone.id);
        Bone updated = bone;
        updated.animationClip = it != animation.boneClips.end() ? it->second : AnimationClip(animation.name);
        scene_.skeleton.setBone(updated);
    }
    for (SceneImage& image : scene_.images) {
        const auto it = animation.imageClips.find(image.id);
        image.animationClip = it != animation.imageClips.end() ? it->second : AnimationClip(animation.name);
    }

    // Restore the constraints to their authored values BEFORE installing
    // the incoming timelines, so a property animated by the outgoing
    // animation but untouched by the incoming one does not keep a stale
    // value (`SceneManager.restoreAllConstraintSetupValues`).
    for (const auto& [constraintID, values] : scene_.constraintSetupValues) {
        if (!constraintKind(scene_.skeleton, constraintID).has_value()) continue;
        applyConstraintSetupValues(scene_.skeleton, constraintID, values);
    }

    scene_.sceneAnimationClip = animation.sceneClip;
    scene_.constraintSetupValues = animation.constraintSetupValues;
    scene_.applyAnimationsNow();
}

} // namespace umeshcore
